contracts/src/internal/HyperdriveBase.sol

```diff
@@ -682,6 +677,152 @@ abstract contract HyperdriveBase is IHyperdriveEvents, HyperdriveStorage {
+    ///      the YieldSpace curve.
+    /// @param _shareReserves The pool's share reserves.
+    /// @param _shareAdjustment The pool's share adjustment.
+    /// @param _bondReserves The pool's bond reserves.
+    /// @param _vaultSharePrice The vault share price.
+    /// @param _longExposure The pool's long exposure.
+    /// @param _checkpointExposure The pool's checkpoint exposure.
+    /// @return The solvency after opening the max long.
+    /// @return A flag indicating whether or not the calculation succeeded.
+    function _calculateSolvencyAfterMaxLongSafe(
+        uint256 _shareReserves,
+        int256 _shareAdjustment,
+        uint256 _bondReserves,
+        uint256 _vaultSharePrice,
+        uint256 _longExposure,
+        int256 _checkpointExposure
+    ) internal view returns (int256, bool) {
+        // Calculate the share payment and bond proceeds of opening the largest
+        // possible long on the YieldSpace curve. This does not include fees.
+        // These calculations fail when the max long is close to zero, and we
+        // ignore these failures since we can proceed with the calculation in
+        // this case.
+        (uint256 effectiveShareReserves, bool success) = HyperdriveMath
+            .calculateEffectiveShareReservesSafe(
+                _shareReserves,
+                _shareAdjustment
+            );
+        if (!success) {
+            return (0, false);
+        }
+        (uint256 maxSharePayment, ) = YieldSpaceMath
+            .calculateMaxBuySharesInSafe(
+                effectiveShareReserves,
+                _bondReserves,
+                ONE - _timeStretch,
+                _vaultSharePrice,
+                _initialVaultSharePrice
+            );
+        (uint256 maxBondProceeds, ) = YieldSpaceMath
+            .calculateBondsOutGivenSharesInDownSafe(
+                effectiveShareReserves,
+                _bondReserves,
+                maxSharePayment,
+                ONE - _timeStretch,
+                _vaultSharePrice,
+                _initialVaultSharePrice
+            );
+
+        // If one of the max share payment or max bond proceeds calculations
+        // fail or return zero, the max long amount is zero plus or minus a few
+        // wei.
+        if (maxSharePayment == 0 || maxBondProceeds == 0) {
+            maxSharePayment = 0;
+            maxBondProceeds = 0;
+        }
+
+        // Apply the fees from opening a long to the max share payment and bond
+        // proceeds. Fees can impact solvency, so it's important to account for
+        // them here.
+        uint256 spotPrice = HyperdriveMath.calculateSpotPrice(
+            effectiveShareReserves,
+            _bondReserves,
+            _initialVaultSharePrice,
+            _timeStretch
+        );
+        (maxSharePayment, maxBondProceeds, ) = _calculateOpenLongFees(
+            maxSharePayment,
+            maxBondProceeds,
+            _vaultSharePrice,
+            spotPrice
+        );
```

Comment:
i'm just curious, couldn't we have used _calculateOpenLong now that we are accounting for fees?

Reply:
We can’t because _calculateOpenLong includes several invariant checks that we don’t want to fail for this calculation (the negative interest checks and the checks within YieldSpaceMath). I could make it a safe function that returns the bytes for an error, which would allow us to reuse the code. I actually kind of like that idea, and we can chat more about it tomorrow